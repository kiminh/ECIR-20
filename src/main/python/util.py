def all_files_in_hdfs_dir(hdfs_path):
    import os
    import re
    import subprocess

    cmd = 'hdfs dfs -ls ' + hdfs_path
    cmd_output = subprocess.check_output(cmd, shell=True)
    ret = []
    for line in cmd_output.decode('utf-8').strip().split('\n'):
        fields = re.split('\s+', line)
        if len(fields) == 8:
            ret += [fields[-1]]

    return ret

