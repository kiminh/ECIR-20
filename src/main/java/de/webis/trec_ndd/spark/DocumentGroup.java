package de.webis.trec_ndd.spark;

import java.io.Serializable;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.TreeSet;
import java.util.stream.Collectors;

import com.fasterxml.jackson.databind.ObjectMapper;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.SneakyThrows;
import lombok.experimental.Accessors;
import scala.Tuple2;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Accessors(chain = true)
@SuppressWarnings("serial")
public class DocumentGroup implements Serializable {
	
	public String hash;
	public ArrayList<String> ids = new ArrayList<String>();
	
	public DocumentGroup (String line) {
		DocumentGroup parsedGroup = createFromJson(line);
		this.hash = parsedGroup.hash;
		this.ids = parsedGroup.ids;
	}
	
	public DocumentGroup (Tuple2<String, Iterable<Tuple2<String, String>>> group) {
		this.hash=group._1;
		group._2.forEach(tuple -> this.ids.add(tuple._2));
	}
	
	@SneakyThrows
	public String toString() {
		return new ObjectMapper().writeValueAsString(this);
	}
	
	@SneakyThrows
	public static List<DocumentGroup> readFromJonLines(Path path) {
		return Files.readAllLines(path).stream()
				.map(s -> createFromJson(s))
				.collect(Collectors.toList());
	}
	
	@SneakyThrows
	public static DocumentGroup createFromJson(String jsonLine) {
		return new ObjectMapper().readValue(jsonLine, DocumentGroup.class);
	}

	public boolean isUsed(HashSet<String> usedDocIDs) {
		for(String id : ids) if(usedDocIDs.contains(id)) return true;
		return false;
	}
	
	public DocumentGroup copy() {
		return new DocumentGroup(this.getHash(), new ArrayList<>(this.getIds()));
	}

	public static List<DocumentGroup> appendRetrievalEquivalentToContentEquivalent(List<DocumentGroup> contentEquivalent, List<DocumentGroup> retrievalEquivalent) {
		List<DocumentGroup> ret = new LinkedList<>(contentEquivalent);
		
		for(DocumentGroup retEq: retrievalEquivalent) {
			boolean shouldAddCompleteRetEq = Boolean.TRUE;
			
			for(DocumentGroup contEq: contentEquivalent) {
				for(String id: retEq.getIds()) {
					if(contEq.getIds().contains(id)) {
						contEq.getIds().addAll(retEq.getIds());
						shouldAddCompleteRetEq = Boolean.FALSE;
						continue;
					}
				}
			}
			
			if(shouldAddCompleteRetEq) {
				ret.add(retEq);
			}
		}
		
		for(DocumentGroup docGroup : ret) {
			docGroup.setIds(new ArrayList<>(new TreeSet<>(docGroup.getIds())));
		}
		
		return ret;
	}
}



