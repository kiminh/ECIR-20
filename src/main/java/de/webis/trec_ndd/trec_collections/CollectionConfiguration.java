package de.webis.trec_ndd.trec_collections;

import java.io.InputStream;
import java.io.Serializable;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import de.webis.trec_ndd.trec_collections.SharedTask.TrecSharedTask;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.SneakyThrows;

public interface CollectionConfiguration extends Serializable {

	public String getPathToCollection();
	public String getCollectionType();
	public String getDocumentGenerator();
	public List<SharedTask> getSharedTasks();

	@SneakyThrows
	public default Set<String> judgedDocumentIds() {
		Set<String> ret = new HashSet<>();
		
		for(String qrelResource : getQrelResourcesOfAllSharedTasks()) {
			InputStream resource = AnseriniCollectionReader.class.getResourceAsStream(qrelResource);
			ret.addAll(new QrelReader().readJudgedDocumentIdsFromQrels(resource));
		}
		
		return ret;
	}
	
	public default HashSet<String> documentIdsInRunFiles() {
		HashSet<String> ret = new HashSet<>();
		
		for(SharedTask sharedTask: getSharedTasks()) {
			ret.addAll(sharedTask.documentIdsInRunFiles());
		}
		
		return ret;
	}
	
	public default List<String> getQrelResourcesOfAllSharedTasks() {
		return getSharedTasks().stream()
				.map(SharedTask::getQrelResource)
				.collect(Collectors.toList());
	}
	
	@Getter
	@AllArgsConstructor
	public static enum TrecCollections implements CollectionConfiguration {
		CLUEWEB09(
			// compressed 4TB
			"/mnt/ceph/storage/corpora/corpora-thirdparty/corpus-clueweb09/",
			"ClueWeb09Collection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(
				TrecSharedTask.WEB_2009,
				TrecSharedTask.WEB_2010,
				TrecSharedTask.WEB_2011,
				TrecSharedTask.WEB_2012,
				TrecSharedTask.SESSION_2010,
				TrecSharedTask.SESSION_2011,
				TrecSharedTask.SESSION_2012
			))
		),
		CLUEWEB12(
			"/mnt/ceph/storage/data-in-progress/kibi9872/clueweb12/",
			"ClueWeb12Collection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(
					TrecSharedTask.WEB_2013,
					TrecSharedTask.WEB_2014,
					TrecSharedTask.SESSION_2013,
					TrecSharedTask.SESSION_2014
			))
		),
		GOV1(
			// compressed 4.6 GB
			"/mnt/nfs/webis20/corpora/corpora-thirdparty/corpora-trec/corpus-trec-web/DOTGOV/",
			"TrecwebCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Collections.emptyList())
		),
		GOV2(
			// compressed 81GB
			"/mnt/nfs/webis20/corpora/corpora-thirdparty/corpora-trec/corpus-trec-web/DOTGOV2/gov2-corpus/",
			"TrecwebCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(
					TrecSharedTask.TERABYTE_2004,
					TrecSharedTask.TERABYTE_2005_ADHOC,
					TrecSharedTask.TERABYTE_2006
			))
		),
		GOV2_MQ(			// compressed 81GB
				"/mnt/ceph/storage/corpora/corpora-thirdparty/corpora-trec/corpus-trec-web/DOTGOV2/gov2-corpus/",
				"TrecwebCollection",
				"JsoupGenerator",
				Collections.unmodifiableList(Arrays.asList(
					TrecSharedTask.MILLION_QUERY_2007,
					TrecSharedTask.MILLION_QUERY_2008
				))
		),
		ROBUST04(
			"/mnt/ceph/storage/data-in-progress/kibi9872/robust-04/",
			"TrecCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(TrecSharedTask.ROBUST_04))
		),
		CORE2018(
			"/mnt/nfs/webis20/data-in-progress/trec-2018-webis/WashingtonPost/WashingtonPost.v2/data/",
			"WashingtonPostCollection",
			"WapoGenerator",
			Collections.unmodifiableList(Arrays.asList(TrecSharedTask.CORE_2018))
		),
		CORE2017(
			"/mnt/nfs/webis20/data-in-progress/trec-2017-webis/data/nyt/",
			"NewYorkTimesCollection",
			"JsoupGenerator",
			Collections.unmodifiableList(Arrays.asList(TrecSharedTask.CORE_2017))
		);
		

		private final String pathToCollection,
							collectionType,
							documentGenerator;

		private final List<SharedTask> sharedTasks;
	}
	
	@Getter
	@AllArgsConstructor
	public static enum OtherCollections implements CollectionConfiguration {
		COMMON_CRAWL_2015_02("CommonCrawlCollection", "JsoupGenerator");
		
		private final String collectionType,
			documentGenerator;

		@Override
		public List<SharedTask> getSharedTasks() {
			return Collections.emptyList();
		}

		@Override
		public String getPathToCollection() {
			return null;
		}
	}
}
