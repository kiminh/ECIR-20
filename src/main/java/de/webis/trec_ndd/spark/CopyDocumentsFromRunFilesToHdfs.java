package de.webis.trec_ndd.spark;

import java.util.List;

import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import de.webis.trec_ndd.spark.SparkBuild8GrammIndex.DocumentSelectionStrategy;
import de.webis.trec_ndd.trec_collections.AnseriniCollectionReader;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import io.anserini.collection.ClueWeb09Collection.Document;
import lombok.Getter;
import net.sourceforge.argparse4j.ArgumentParsers;
import net.sourceforge.argparse4j.inf.ArgumentParser;
import net.sourceforge.argparse4j.inf.Namespace;

public class CopyDocumentsFromRunFilesToHdfs implements SparkArguments {
	
	@Getter
	private final Namespace parsedArgs;
	
	private CopyDocumentsFromRunFilesToHdfs(String[] args) {
		this.parsedArgs = parseArguments(args);
	}
	
	@Override
	public void run() {
		DocumentSelectionStrategy selectionStrategy = documentSelection();
		AnseriniCollectionReader<Document> acr = new AnseriniCollectionReader<Document>(collection());
		List<String> segmentPaths = acr.segmentPaths();

		try (JavaSparkContext context = context()) {
			JavaRDD<CollectionDocument> documents = context.parallelize(segmentPaths)
				.flatMap(s -> selectionStrategy.getDocumentsInSegmentPath().apply(acr).apply(s));
			
			documents.saveAsTextFile(jobName());
		}
	}
	
	public static void main(String[] args) {
		new CopyDocumentsFromRunFilesToHdfs(args).run();
	}
	
	private Namespace parseArguments(String[] args) {
		ArgumentParser parser = ArgumentParsers.newFor("CopyDocumentsToHdfs")
			.build()
			.defaultHelp(true)
			.description("Copy all selected documents to hdfs for faster access.");

		addCollectionToArgparser(parser);
		addDocumentSelectionToArgparser(parser);

		return parser.parseArgsOrFail(args);
	}
	
	public String jobName() {
		return jobName(collectionName(), documentSelection());
	}
	
	public static String jobName(String collectionName, DocumentSelectionStrategy documentSelection) {
		if(DocumentSelectionStrategy.ALL.equals(documentSelection)) {
			throw new RuntimeException("It is too expensive to use document-selection: " + documentSelection.name());
		}
		
		return "trec-docs-in-" + documentSelection.name().toLowerCase() + "-for-" + collectionName.toLowerCase();
	}
}
