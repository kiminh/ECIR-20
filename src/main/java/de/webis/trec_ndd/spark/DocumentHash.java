package de.webis.trec_ndd.spark;

import java.io.Serializable;
import java.util.HashSet;
import java.util.List;

import org.codehaus.jackson.map.ObjectMapper;

import de.webis.trec_ndd.similarity.MD5;
import de.webis.trec_ndd.similarity.TextProfileSignatureSimilarity;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import de.webis.trec_ndd.util.NGramms;
import de.webis.trec_ndd.util.NGramms.Word8Gramm;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.SneakyThrows;
import lombok.experimental.Accessors;

@Data
@NoArgsConstructor
@Accessors(chain = true)
@SuppressWarnings("serial")
public class DocumentHash implements Serializable {
	private String id;
	private String textProfileSignature;
	private String md5;
	private String fullyTextProfileSignature;
	private String fullyCanonicalizedMd5;
	private long documentLength;
	private long fullyCanonicalizedWord8GrammCount;
	private long fullyCanonicalizedWord8GrammSetSize;
	
	public DocumentHash (CollectionDocument doc) {
		this.id=doc.getId();
		this.md5=MD5.md5hash(doc.getContent());
		this.textProfileSignature=TextProfileSignatureSimilarity.textProfileSignatureString(doc.getContent());
		this.documentLength=doc.getContent().length();
		this.fullyTextProfileSignature = TextProfileSignatureSimilarity.textProfileSignatureString(doc.getFullyCanonicalizedContent());
		this.fullyCanonicalizedMd5 = MD5.md5hash(doc.getFullyCanonicalizedContent());
		List<Word8Gramm> word8Gramms = NGramms.build8Gramms(doc.getFullyCanonicalizedContent());
		
		this.fullyCanonicalizedWord8GrammCount = word8Gramms.size();
		this.fullyCanonicalizedWord8GrammSetSize = new HashSet<>(word8Gramms).size();
	}
	
	@SneakyThrows
	public String toString() {
		return new ObjectMapper().writeValueAsString(this);
	}
	
	@SneakyThrows
	public static DocumentHash fromString(String str) {
		return new ObjectMapper().readValue(str, DocumentHash.class);
	}
}
