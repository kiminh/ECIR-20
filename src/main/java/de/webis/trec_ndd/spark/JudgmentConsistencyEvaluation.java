package de.webis.trec_ndd.spark;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import org.apache.spark.api.java.JavaSparkContext;
import org.codehaus.jackson.map.ObjectMapper;

import com.google.common.collect.ImmutableMap;

import static de.webis.trec_ndd.spark.SparkDeduplicateAndEvaluate.readGroupsUsedInRunFiles;

import de.webis.trec_ndd.trec_collections.SharedTask;
import de.webis.trec_ndd.trec_collections.SharedTask.DocumentJudgments;
import lombok.Builder;
import lombok.Data;
import lombok.Getter;
import lombok.SneakyThrows;
import de.webis.trec_ndd.trec_collections.CollectionConfiguration.TrecCollections;
import net.sourceforge.argparse4j.ArgumentParsers;
import net.sourceforge.argparse4j.inf.ArgumentParser;
import net.sourceforge.argparse4j.inf.Namespace;

public class JudgmentConsistencyEvaluation implements SparkArguments {
	
	@Getter
	private final Namespace parsedArgs;
	
	public JudgmentConsistencyEvaluation(String[] args) {
		this.parsedArgs = parseArguments(args);
	}
	
	@Override
	public void run() {
		try (JavaSparkContext sc = context()) {
			for(TrecCollections collection: collections()) {
				String retrievalEquivalentDir = SparkGroupByFingerprint.resultDir(collection, documentSelection(), documentSimilarity());
				List<DocumentGroup> retrievalEquivalent = readGroupsUsedInRunFiles(sc, retrievalEquivalentDir, collection); 
				String contentEquivalentDir = SparkGroupByFingerprint.resultDir(collection, documentSelection(), "s3-similarity-connected-component-" + s3Threshold());
				List<DocumentGroup> contentEquivalent = readGroupsUsedInRunFiles(sc, contentEquivalentDir, collection);
				List<DocumentGroup> contentEquivalentWithoutRetrievalEquivalentOnly = new ArrayList<>(contentEquivalent.stream().map(DocumentGroup::copy).collect(Collectors.toList()));
				contentEquivalent = DocumentGroup.appendRetrievalEquivalentToContentEquivalent(contentEquivalent, retrievalEquivalent);
				
				DocumentGroups documentGroups = DocumentGroups.builder()
						.contentEquivalent(contentEquivalent)
						.retrievalEquivalent(retrievalEquivalent)
						.contentEquivalentWithoutRetrievalEquivalent(contentEquivalentWithoutRetrievalEquivalentOnly)
						.build();
				
				for(SharedTask task : collection.getSharedTasks()) {
					report(documentGroups, task);
				}
			}
		}
	}
	
	public static void main(String[] args) {
		new JudgmentConsistencyEvaluation(args).run();
	}
	
	@SneakyThrows
	private static void report(DocumentGroups docGroups, SharedTask task) {
		DocumentJudgments docJudgments = task.documentJudgments();
		
		Map<String, Object> report = ImmutableMap.<String, Object>builder()
				.put("SharedTask", task.name())
				.put("GroupSize", docGroups.getContentEquivalent().size())
				.put("InconsistentGroups", docGroups.inconsistentGroups(docJudgments))
				.put("InconsistentGroups(WithoutUnlabeled)", docGroups.inconsistentGroupsWithoutUnlabeled(docJudgments))
				.put("InconsistentGroups(WithoutUnlabeledAndRetrievalEquivalent)", docGroups.inconsistentGroupsWithoutUnlabeledWithoutRetrievalEquivalent(docJudgments))
				.put("ConsistentGroups", docGroups.consistentGroups(docJudgments))
				.put("JudgmentCount", docJudgments.judgmentCount())
				.put("DocumentsToJudgeAgain", docGroups.documentsToJudgeAgain(docJudgments))
				.put("TopicDetails", docGroups.topicDetails(docJudgments))
				.build();
		
		System.out.println(new ObjectMapper().writeValueAsString(report));
	}

	private Namespace parseArguments(String[] args) {
		ArgumentParser parser = ArgumentParsers.newFor("JudgmentConsistencyEvaluation")
				.build()
				.defaultHelp(true)
				.description("Evaluate inconsistencies caused by near-duplicates in the QRELs.");
		
		addCollectionsToArgparser(parser);
		addSimilarityToArgparser(parser);
		addDocumentSelectionToArgparser(parser);
		addS3Threshold(parser);
		
		return parser.parseArgsOrFail(args);
	}
	
	@Data
	@Builder
	public static class DocumentGroups {
		private List<DocumentGroup> contentEquivalent,
									retrievalEquivalent,
									contentEquivalentWithoutRetrievalEquivalent;
		
		public int inconsistentGroups(DocumentJudgments docJudgments) {
			return contentEquivalent.parallelStream()
				.mapToInt(group -> docJudgments.groupHasInconsistency(group) ? 1 : 0)
				.sum();
		}
		
		public int inconsistentGroupsWithoutUnlabeled(DocumentJudgments docJudgments) {
			return contentEquivalent.parallelStream()
				.mapToInt(group -> docJudgments.groupHasInconsistencyWithoutUnlabeled(group) ? 1 : 0)
				.sum();
		}
		
		public Object inconsistentGroupsWithoutUnlabeledWithoutRetrievalEquivalent(DocumentJudgments docJudgments) {
			return contentEquivalentWithoutRetrievalEquivalent.parallelStream()
				.mapToInt(group -> docJudgments.groupHasInconsistencyWithoutUnlabeled(group) ? 1 : 0)
				.sum();
		}

		public int consistentGroups(DocumentJudgments docJudgments) {
			return contentEquivalent.parallelStream()
				.mapToInt(group -> docJudgments.groupHasInconsistency(group) ? 0 : 1)
				.sum();
		}
		
		public int documentsToJudgeAgain(DocumentJudgments docJudgments) {
			return contentEquivalent.parallelStream()
				.mapToInt(docJudgments::documentsToJudgeAgain)
				.sum();
		}

		public Map<String, Map<String, Integer>> topicDetails(DocumentJudgments docJudgments) {
			return docJudgments.topics().stream()
				.collect(Collectors.toMap(i -> i, i -> topicDetails(i, docJudgments)));
		}
		
		private Map<String, Integer> topicDetails(String topic, DocumentJudgments docJudgments) {
			Set<String> relevantJudgments = new HashSet<>(docJudgments.getRelevantDocuments(topic));
			Set<String> irrelevantJudgments = new HashSet<>(docJudgments.getIrrelevantDocuments(topic));
			Set<String> retrievalEquivalent = getRetrievalEquivalent().stream()
				.flatMap(i -> i.ids.stream())
				.collect(Collectors.toSet());
			Set<String> contentEquivalent = getContentEquivalent().stream()
					.flatMap(i -> i.ids.stream())
					.collect(Collectors.toSet());
			
			failOnDocumentsThatAreRetrievalEquivalentButNotContentEquivalent(retrievalEquivalent, contentEquivalent, topic);
			
			return ImmutableMap.<String, Integer>builder()
				.put("relevantJudgments", relevantJudgments.size())
				.put("retrievalEquivalentGroupsInRelevantJudgments", groupsInIntersection(relevantJudgments, getRetrievalEquivalent()))
				.put("contentEquivalentGroupsInRelevantJudgments", groupsInIntersection(relevantJudgments, getContentEquivalent()))
				.put("retrievalEquivalentDocsInRelevantJudgments", intersection(retrievalEquivalent, relevantJudgments))
				.put("contentEquivalentDocsInRelevantJudgments", intersection(contentEquivalent, relevantJudgments))
				.put("irrelevantJudgments", irrelevantJudgments.size())
				.put("retrievalEquivalentDocsInIrrelevantJudgments", intersection(retrievalEquivalent, irrelevantJudgments))
				.put("contentEquivalentDocsInIrrelevantJudgments", intersection(contentEquivalent, irrelevantJudgments))
				.put("retrievalEquivalentGroupsInIrrelevantJudgments", groupsInIntersection(irrelevantJudgments, getRetrievalEquivalent()))
				.put("contentEquivalentGroupsInIrrelevantJudgments", groupsInIntersection(irrelevantJudgments, getContentEquivalent()))
				.build();
		}
		
		private static void failOnDocumentsThatAreRetrievalEquivalentButNotContentEquivalent(Set<String> retrievalEquivalent, Set<String> contentEquivalent, String topic) {
			List<String> notContentEquivalent = new ArrayList<>();
			
			for(String retrievalEq : retrievalEquivalent) {
				if(!contentEquivalent.contains(retrievalEq)) {
					notContentEquivalent.add(retrievalEq);
				}
			}
			
			if(notContentEquivalent.size() > 0) {
				throw new RuntimeException("Unexpected Behaviour for topic '" + topic + "': Documents are retrieval-equivalent but not content-equivalent: "+ notContentEquivalent);
			}
		}
		
		private int groupsInIntersection(Set<String> a, List<DocumentGroup> b) {
			int ret = 0;
			
			for(DocumentGroup docGroup: b) {
				if(docGroup.ids.stream().anyMatch(i -> a.contains(i))) {
					ret += 1;
				}
			}
			
			return ret;
		}
		
		private <T> int intersection(Set<T> a, Set<T> b) {
			Set<T> ret = new HashSet<T>(a);
			ret.retainAll(b);
			
			return ret.size();
		}
	}

	@Override
	public String jobName() {
		return "trec-qrel-consistency";
	}
}
