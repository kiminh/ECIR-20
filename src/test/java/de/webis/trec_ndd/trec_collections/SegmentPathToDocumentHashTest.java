package de.webis.trec_ndd.trec_collections;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import org.approvaltests.Approvals;
import org.junit.Test;

import com.google.common.collect.Lists;

import de.webis.trec_ndd.spark.DocumentHash;
import de.webis.trec_ndd.trec_collections.CollectionConfiguration.TrecCollections;

public class SegmentPathToDocumentHashTest {

	@Test
	public void approveTransformationForRobustCollectionWithSingleSegment() {
		CollectionConfiguration conf = SegmentPathToDocumentTextTest.collectionConfiguration("src/test/resources/data/robust", TrecCollections.ROBUST_TBD);
		AnseriniCollectionReader<?> collectionReader = new AnseriniCollectionReader<>(conf);
		List<DocumentHash> collectionDocuments = mapAllSegmentPathsToDocumentRepresentation(collectionReader);
		
		Approvals.verifyAsJson(collectionDocuments);
	}
	
	@Test
	public void approveTransformationForClueweb09CollectionWithSingleSegment() {
		CollectionConfiguration conf = SegmentPathToDocumentTextTest.collectionConfiguration("src/test/resources/data/clueweb09-sample", TrecCollections.CLUEWEB09);
		AnseriniCollectionReader<?> collectionReader = new AnseriniCollectionReader<>(conf);
		List<DocumentHash> collectionDocuments = mapAllSegmentPathsToDocumentRepresentation(collectionReader);
		
		Approvals.verifyAsJson(collectionDocuments);
	}
	
	@Test
	public void approveTransformationForClueweb09CollectionWithMultipleSegments() {
		CollectionConfiguration conf = SegmentPathToDocumentTextTest.collectionConfiguration("src/test/resources/data/clueweb09-sample-2", TrecCollections.CLUEWEB09);
		AnseriniCollectionReader<?> collectionReader = new AnseriniCollectionReader<>(conf);
		List<DocumentHash> collectionDocuments = mapAllSegmentPathsToDocumentRepresentation(collectionReader);
		
		Approvals.verifyAsJson(collectionDocuments);
	}
	
	@Test
	public void approveTransformationForGOV1CollectionWithMultipleSegments() {
		CollectionConfiguration conf = SegmentPathToDocumentTextTest.collectionConfiguration("src/test/resources/data/gov1-sample", TrecCollections.GOV1);
		AnseriniCollectionReader<?> collectionReader = new AnseriniCollectionReader<>(conf);
		List<DocumentHash> collectionDocuments = mapAllSegmentPathsToDocumentRepresentation(collectionReader);
		
		Approvals.verifyAsJson(collectionDocuments);
	}
	
	@Test
	public void approveTransformationForGOV2CollectionWithMultipleSegments() {
		CollectionConfiguration conf = SegmentPathToDocumentTextTest.collectionConfiguration("src/test/resources/data/gov2-sample", TrecCollections.GOV2);
		AnseriniCollectionReader<?> collectionReader = new AnseriniCollectionReader<>(conf);
		List<DocumentHash> collectionDocuments = mapAllSegmentPathsToDocumentRepresentation(collectionReader);
		
		Approvals.verifyAsJson(collectionDocuments);
	}
	
	@Test
	public void approveTransformationForCore2018CollectionWithSingleSegments() {
		CollectionConfiguration conf = SegmentPathToDocumentTextTest.collectionConfiguration("src/test/resources/data/core-2018-sample", TrecCollections.CORE2018);
		AnseriniCollectionReader<?> collectionReader = new AnseriniCollectionReader<>(conf);
		List<DocumentHash> collectionDocuments = mapAllSegmentPathsToDocumentRepresentation(collectionReader);
		
		Approvals.verifyAsJson(collectionDocuments);
	}
	
	@Test
	public void approveTransformationForCore2017CollectionWithMultipleSegments() {
		CollectionConfiguration conf = SegmentPathToDocumentTextTest.collectionConfiguration("src/test/resources/data/core-2017-sample", TrecCollections.CORE2017);
		AnseriniCollectionReader<?> collectionReader = new AnseriniCollectionReader<>(conf);
		List<DocumentHash> collectionDocuments = mapAllSegmentPathsToDocumentRepresentation(collectionReader);
		Collections.sort(collectionDocuments, (a,b) -> a.getId().compareTo(b.getId()));
		
		Approvals.verifyAsJson(collectionDocuments);
	}
	
	private static List<DocumentHash> mapAllSegmentPathsToDocumentRepresentation(AnseriniCollectionReader<?> collectionReader) {
		return collectionReader.segmentPaths().stream()
				.map(i -> collectionReader.documentHashIterator(i))
				.map(Lists::newArrayList)
				.flatMap(List::stream)
				.collect(Collectors.toList());
	}
}
