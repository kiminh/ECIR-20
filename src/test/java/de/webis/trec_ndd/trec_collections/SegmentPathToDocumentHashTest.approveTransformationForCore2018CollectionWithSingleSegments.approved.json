[
  {
    "id": "fffbc452-b422-11e2-9a98-4be1688d7d84",
    "textProfileSignature": "672a6cc5f7e1f66f250a38414b95db5b",
    "md5": "b6bbba0bacf3f58104ee29e94b0eeebc",
    "fullyTextProfileSignature": "a6582fc82b0c31bbf8ae018282ef0b41",
    "fullyCanonicalizedMd5": "b4bde8b324baf7f457ebac817f8dbf93",
    "documentLength": 3957,
    "fullyCanonicalizedWord8GrammCount": 460,
    "fullyCanonicalizedWord8GrammSetSize": 460
  },
  {
    "id": "b2e89334-33f9-11e1-825f-dabc29fd7071",
    "textProfileSignature": "d8f6f7e167203ee8f2752b3acacfa70d",
    "md5": "bb8ca0ca895d614b731b8b702e057040",
    "fullyTextProfileSignature": "b0bc755749c5bee91c29e5da5e66b27e",
    "fullyCanonicalizedMd5": "d8fac046a74f29876c1c0d168de4173f",
    "documentLength": 4791,
    "fullyCanonicalizedWord8GrammCount": 607,
    "fullyCanonicalizedWord8GrammSetSize": 607
  },
  {
    "id": "749ec5b2-32f5-11e1-825f-dabc29fd7071",
    "textProfileSignature": "1e351d52530158b03699249c6641a759",
    "md5": "6e2487d95bf84732893774e75d49f607",
    "fullyTextProfileSignature": "2aad319b01f1d39ad148f133ef90236c",
    "fullyCanonicalizedMd5": "670ca183f9a67aff03b33e0255d4d6d2",
    "documentLength": 4017,
    "fullyCanonicalizedWord8GrammCount": 484,
    "fullyCanonicalizedWord8GrammSetSize": 484
  }
]